From: Borislav Petkov <bp@suse.de>
Date: Wed, 10 Jan 2018 00:10:42 +0100
Subject: xen/x86/CPU: Sync CPU feature flags late
Patch-mainline: Never, SUSE-Xen specific
References: bsc#1075994 bsc#1075091

This is for the case where we need to set feature flags late, like, for
example, after late microcode patch has been loaded which has enabled
new CPUID bits.

This has no effect on alternatives patching.

Signed-off-by: Borislav Petkov <bp@suse.de>
Cc: Peter Zijlstra <peterz@infradead.org>
Cc: x86@kernel.org
Automatically created from "patches.suse/33-x86-CPU-Sync-CPU-feature-flags-late.patch" by xen-port-patches.py

--- a/arch/x86/kernel/cpu/common-xen.c
+++ b/arch/x86/kernel/cpu/common-xen.c
@@ -660,6 +660,25 @@ static void apply_forced_caps(struct cpu
 	}
 }
 
+/*
+ * This late synchronization of CPU caps has no effect on alternatives patching
+ * but updates the visible feature bits per CPU.
+ *
+ * Callers need to disable CPU hotplug around it.
+ */
+void cpu_caps_sync_late(void)
+{
+	int cpu;
+
+
+	for_each_online_cpu(cpu) {
+		struct cpuinfo_x86 *c = &cpu_data(cpu);
+
+		apply_forced_caps(c);
+	}
+}
+EXPORT_SYMBOL_GPL(cpu_caps_sync_late);
+
 void __cpuinit get_cpu_cap(struct cpuinfo_x86 *c)
 {
 	u32 tfms, xlvl;
